packages/mcdoc/src/validator/index.ts

```diff
@@ -0,0 +1,896 @@
+					f.key.value.value === lookup
+				)
+				if (value) {
+					values.push(value.type)
+				} else {
+					// fallback case
+					values = child.fields.filter(f => f.kind === 'pair').map(f =>
+						f.type
+					)
+					break
+				}
+			}
+			return simplify(
+				node,
+				{ kind: 'union', members: values },
+				inferredType,
+				options,
+				parents,
+			)
+		case 'union':
+			const members: SimplifiedMcdocTypeNoUnion[] = []
+			for (const member of typeDef.members) {
+				const simplified = simplify(
+					node,
+					member,
+					inferredType,
+					options,
+					parents,
+				)
+
+				if (simplified.kind === 'union') {
+					members.push(...simplified.members)
+				} else {
+					members.push(simplified)
+				}
+			}
+			if (members.length === 1) {
+				return members[0]
+			}
+			return { kind: 'union', members: members }
+		case 'struct':
+			const fields: SimplifiedStructTypePairField[] = []
+			for (const field of typeDef.fields) {
+				if (field.kind === 'pair') {
+					// Don't simplify the value here. We need to have the correct `node` and `parents`, which we
+					// cannot deterministically find for non-string keys.
+					// Instead, this method will be called by every struct child by the outer checking method.
+					const key = typeof field.key === 'string'
+						? field.key
+						: simplify(node, field.key, inferredType, options, parents)
+					if (typeof key !== 'string' && key.kind === 'union') {
+						for (const subKey of key.members) {
+							fields.push({
+								...field,
+								key: simplifyKey(subKey),
+							})
+						}
+					} else {
+						fields.push({
+							...field,
+							key: simplifyKey(key),
+						})
+					}
+				} else {
+					const simplified = simplify(
+						node,
+						field.type,
+						inferredType,
+						options,
+						parents,
+					)
+
+					if (simplified.kind === 'struct') {
+						fields.push(...simplified.fields)
+					}
+				}
+			}
+			return {
+				kind: 'struct',
+				fields: fields.filter((f, i) =>
+					i <=
+						fields.findLastIndex(of =>
+							isAssignable(
+								f.key,
+								of.key,
+								options.context,
+								options.isEquivalent,
+							)
+						)
+				),
+			}
+		case 'enum':
+			return { ...typeDef, enumKind: typeDef.enumKind ?? 'int' }
+		case 'concrete': // TODO
+		case 'template': // TODO
+			return { kind: 'union', members: [] }
+		default:
+			return typeDef
+	}
+}
+function simplifyKey(
+	keyDef: SimplifiedMcdocTypeNoUnion | string,
+): SimplifiedMcdocTypeNoUnion {
+	if (typeof keyDef == 'string') {
+		return { kind: 'literal', value: { kind: 'string', value: keyDef } }
+	}
+	return keyDef
+}
+
+function getValueType(
+	type: SimplifiedMcdocType,
+): Exclude<SimplifiedMcdocTypeNoUnion, LiteralType | EnumType>[] {
+	switch (type.kind) {
+		case 'literal':
+			return [{ kind: type.value.kind }]
+		case 'enum':
+			return [{ kind: type.enumKind }]
+		case 'union':
+			return type.members.flatMap(m => getValueType(m))
```

Comment:
Should this remove duplicates?

Reply:
Yes, this is out of scope though as I would need to fix a rather difficult issue in the isAssignable / validate logic.

The issue is related to recursive type definitions, were e.g. a text component type has another text component type as a child. In this case I would recursively resolve the reference and keep checking if they are equal.

As a simple hack, you could add a check to see if both are a reference and they are equal, this would break if you have two different recursive but in principle assignable definitions though.

Reply:
I created an issue for the problem I described #1228